Spigot-Server-Patches/0402-implement-optional-per-player-mob-spawns.patch

```diff
@@ -0,0 +1,244 @@
+                                                         ServerInternalException.reportInternalException(exception); // Paper
+-                                                        return;
++                                                        return mobsSpawned; // Paper
+                                                     }
+ 
+                                                     entityinsentient.setPositionRotation((double) f, (double) k, (double) f1, world.random.nextFloat() * 360.0F, 0.0F);
+@@ -114,10 +122,17 @@ public final class SpawnerCreature {
+                                                         if (world.addEntity(entityinsentient, SpawnReason.NATURAL)) {
+                                                             ++i;
+                                                             ++i2;
++                                                            // Paper start - stop when limit is reached
++                                                            mobsSpawned.add(entityinsentient);
++                                                        }
++
++                                                        if (mobsSpawned.size() >= maxSpawns){
++                                                            return mobsSpawned; // Paper
+                                                         }
++                                                        // Paper end
+                                                         // CraftBukkit end
+                                                         if (i >= entityinsentient.dC()) {
+-                                                            return;
++                                                            return mobsSpawned; // Paper
+                                                         }
+ 
+                                                         if (entityinsentient.c(i2)) {
+@@ -142,6 +157,7 @@ public final class SpawnerCreature {
+ 
+             }
+         }
++        return mobsSpawned; // Paper
+     }
+ 
+     @Nullable
+diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
+index 135ec94c..638708f1 100644
+--- a/src/main/java/net/minecraft/server/WorldServer.java
++++ b/src/main/java/net/minecraft/server/WorldServer.java
+@@ -17,6 +17,8 @@ import it.unimi.dsi.fastutil.objects.Object2IntMap;
+ import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
+ import it.unimi.dsi.fastutil.objects.ObjectIterator;
+ import it.unimi.dsi.fastutil.objects.ObjectLinkedOpenHashSet;
++import java.util.ArrayList; // Paper
++import java.util.HashMap; // Paper
+ import java.util.Iterator;
+ import java.util.List;
+ import java.util.Map;
+@@ -67,6 +69,7 @@ public class WorldServer extends World {
+     private boolean ticking;
+     @Nullable
+     private final MobSpawnerTrader mobSpawnerTrader;
++    private Map<EntityPlayer, Map<EnumCreatureType, Integer>> playerMobTypeMap; // Paper
```

Comment:
This is a memory leak, you have to remove the entry of the player quits or changes the world, or use something like the UUID of the player.